src/os/bluestore/bluestore_types.cc

```diff
@@ -335,6 +335,241 @@ ostream& operator<<(ostream& out, const bluestore_extent_ref_map_t& m)
   return out;
 }
 
+// bluestore_blob_use_tracker_t
+
+void bluestore_blob_use_tracker_t::allocate()
+{
+  assert(num_au != 0);
+  bytes_per_au = new uint32_t[num_au];
+}
+
+void bluestore_blob_use_tracker_t::fall_back_to_partial(uint32_t _num_au, uint32_t _au_size)
+{
+  assert(num_au == 0 && au_size == 0);
+  assert(_num_au != 0 && _au_size != 0);
+  assert(_num_au * _au_size >= total_bytes);
+
+  uint32_t total = total_bytes;
+  num_au = _num_au;
+  au_size = _au_size;
+  allocate();
+  for (size_t i = 0; i < num_au; ++i) {
+    bytes_per_au[i] = min(total, au_size);
+    total -= bytes_per_au[i];
+  }
+  assert(total == 0);
+}
+
+void bluestore_blob_use_tracker_t::get(
+  uint32_t offset, uint32_t length,
+  std::function<void(uint32_t*, uint32_t*)> fallback_cb)
+{
+  if (!num_au) {
+    if (offset == 0 && total_bytes == 0) {
+      total_bytes = length;
+      return;
+    }
+    uint32_t _au_size = 0, _num_au = 0;
+    fallback_cb(&_num_au, &_au_size);
+    if (_num_au && _au_size) { // we might be forbidden to fallback to partial
+      fall_back_to_partial(_num_au, _au_size);
+    } else {
+      total_bytes += length;
+    }
+  } 
+  if (num_au) {
+    assert(au_size != 0);
+    auto end = offset + length;
+
+    while (offset < end) {
+      auto phase = offset % au_size;
+      bytes_per_au[offset / au_size] += 
+	MIN(au_size - phase, end - offset);
+      offset += (phase ? au_size - phase : au_size);
+    }
+  }
+}
+
+bool bluestore_blob_use_tracker_t::put(
+  uint32_t offset, uint32_t length,
+  std::function<void(uint32_t*, uint32_t*)> fallback_cb,
+  set<uint32_t> *release_units)
+{
+  if (release_units) {
+    release_units->clear();
+  }
+  if (!num_au) {
+    if (total_bytes == length) {
+      total_bytes = 0;
+      return true;
+    }
+    uint32_t _au_size = 0, _num_au = 0;
+    fallback_cb(&_num_au, &_au_size);
+    if (_num_au && _au_size) { // we might be forbidden to fallback to partial
+      fall_back_to_partial(_num_au, _au_size);
+    } else {
+      total_bytes -= length;
+    }
+  }
+  bool maybe_empty = true;
+  if (num_au) {
+    assert(au_size != 0);
+    auto end = offset + length;
+    while (offset < end) {
+      auto phase = offset % au_size;
+      size_t pos = offset / au_size;
+      auto diff = MIN(au_size - phase, end - offset);
+      assert(diff <= bytes_per_au[pos]);
+      bytes_per_au[pos] -= diff;
+      offset += (phase ? au_size - phase : au_size);
+      if (bytes_per_au[pos] == 0) {
+	if (release_units) {
+	  release_units->insert(pos * au_size);
```

Comment:
Reworked to return logical bluestore_pextent_t but modified put_ref totally anyway. The rationales are:
1) reduced complexity - original implementation had O(p*r) vs. O(p+r) for the new one. Where p - blob pextent vector size, r - release vector size.
2) original implementation still needed update since it uses ref_map.